ocrd_kraken/segment.py

```diff
@@ -1,62 +1,111 @@
+            from kraken.blla import segment as blla_segment
+            log.info("Using blla segmenter")
+            blla_model_fname = self.resolve_resource(self.parameter['blla_model'])
+            kwargs['model'] = TorchVGSLModel.load_model(blla_model_fname)
+            kwargs['device'] = self.parameter['device']
+            segment = blla_segment
 
-            log.info('Segmenting')
-            log.info('Params %s', self.parameter)
-            res = segment(
-                im,
-                self.parameter['text_direction'],
-                self.parameter['scale'],
-                self.parameter['maxcolseps'],
-                self.parameter['black_colseps']
-            )
-            if self.parameter['script_detect']:
-                res = detect_scripts(im, res)
-
-            dummyRegion = TextRegionType()
-            pcgts.get_Page().add_TextRegion(dummyRegion)
-            #  print(res)
-            for lineno, box in enumerate(res['boxes']):
-                textline = TextLineType(
-                    id=concat_padded("line", lineno),
-                    Coords=CoordsType(points=points_from_x0y0x1y1(box))
-                )
-                dummyRegion.add_TextLine(textline)
-            ID = concat_padded(self.output_file_grp, n)
+        for (n, input_file) in enumerate(self.input_files):
+            page_id = input_file.pageId or input_file.ID
+            log.info("INPUT FILE %i / %s of %s", n, page_id, len(self.input_files))
+            pcgts = page_from_file(self.workspace.download_file(input_file))
+            self.add_metadata(pcgts)
+            page = pcgts.get_Page()
+            page_image, _, _ = self.workspace.image_from_page(page, page_id, feature_selector="binarized")
+            log.info('Segmenting with %s segmenter' % ('legacy' if use_legacy else 'blla'))
+            res = segment(page_image, **kwargs)
+            log.info("Finished segmentation, serializing")
+            if use_legacy:
+                print(res)
+                for idx_line, line_x0y0x1y1 in enumerate(res['boxes']):
+                    region_elem = TextRegionType(
+                        id='region_line_%s' % (idx_line + 1),
+                        Coords=CoordsType(points=points_from_x0y0x1y1(line_x0y0x1y1)))
+                    region_elem.add_TextLine(TextLineType(
+                        id='region_line_%s_line' % (idx_line + 1),
+                        Coords=CoordsType(points=points_from_x0y0x1y1(line_x0y0x1y1))))
+                    page.add_TextRegion(region_elem)
+            else:
+                handled_lines = {}
+                for idx_region, region_polygon_ in enumerate(res['regions']['text']):
+                    region_elem = TextRegionType(
+                            id='region_%s' % (idx_region + 1),
+                            Coords=CoordsType(points=points_from_polygon(region_polygon_)))
+                    region_polygon = geom.Polygon(region_polygon_)
```

Comment:
Using (low-level) Shapely like that is too fragile I'm afraid. I don't know how well-behaved the blla results are, but you risk getting invalid paths which Shapely cannot operate on. Have a look at ocrd_tesserocr's `make_valid` for reference.

Moreover, even if they are valid, the precision of the neural output will very likely be no more than 1px, whereas Shapely operates on exact real-valued objects. So you might need to use `region_polygon.buffer(1)` to avoid missing lines whose boundary touches the region boundary. 

Also, there's `shapely.prepared.prep` for repeated operations on the same polygons (shifting the time-space tradeoff).

Reply:
cf. 232a055